### PR TITLE
Handle SIGTERM in temporal worker

### PR DESCRIPTION
**Pull request by Aayyush**
In this PR, we configure the `WorkerStopTimeout` to 60 minutes which is equal to the termination grace period configured for our legacy worker. 

The go routine that runs the temporal worker closes the job output channel when exiting which exits the for loop in `streamHandler.Handle()` and triggers clean up to close any active receivers and persists all jobs in memory.

### CODE REVIEW

File: server/logging/logger.go [hunk 1]
```diff
@@ -59,6 +65,7 @@ func NewLoggerFromLevel(lvl LogLevel) (*logger, error) {
 
 	return &logger{
 		LoggerFacade: ctxLogger,
+		Closer:       &closer{},
```
Comment (samrabelachew) [inline server/logging/logger.go:68]: The closer should be the underlying structuredLogger, so that we can flush out the logs with the sync fxn

File: server/neptune/temporalworker/server.go [hunk 2]
```diff
@@ -154,9 +159,14 @@ func (s Server) Start() error {
 
 	go func() {
 		defer wg.Done()
+
+		// Close job stream when temporalworker exits to allow gracefully shutting downt the stream handler
+		defer close(s.JobStream)
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:164]: Why does this need to be in this func() and not after <- stop?

File: server/neptune/temporalworker/server.go [hunk 3]
```diff
@@ -185,9 +195,17 @@ func (s Server) Start() error {
...
+		// Returns when the job output chan is closed which happens only after the temporal worker
+		// has gracefully shut down
 		s.JobStreamHandler.Handle()
+
+		// On cleanup, stream handler closes all active receivers and persists jobs in memory
+		ctx, cancel := context.WithTimeout(context.Background(), StreamHandlerTimeout)
+		defer cancel()
+		if err := s.JobStreamHandler.CleanUp(ctx); err != nil {
+			s.Logger.Error(err.Error())
+		}
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:208]: Why is this here? this function doesn't use a wg so its not guaranteed this will be called, also there might be a race with the stats/logger closing and this.

I would move this after the stop

Comment (Aayyush) [reply server/neptune/temporalworker/server.go:208]: Ahh true!

File: server/neptune/temporalworker/server.go [hunk 4]
```diff
@@ -154,9 +159,14 @@ func (s Server) Start() error {
 
 	go func() {
 		defer wg.Done()
+
+		// Close job stream when temporalworker exits to allow gracefully shutting downt the stream handler
```
Comment (samrabelachew) [inline server/neptune/temporalworker/server.go:163]: down

File: server/neptune/temporalworker/server.go [hunk 5]
```diff
@@ -35,6 +35,8 @@ const (
 	AtlantisNamespace        = "atlantis"
 	DeployTaskqueue          = "deploy"
 	ProjectJobsViewRouteName = "project-jobs-detail"
+	TemporalWorkerTimeout    = 10 * time.Second
```
[the remark this answers is not in the document]
Comment (samrabelachew) [reply server/neptune/temporalworker/server.go:38]: bump this^

File: server/neptune/temporalworker/job/store.go [hunk 6]
```diff
@@ -170,11 +183,31 @@ func (s *StorageBackendJobStore) Close(ctx context.Context, jobID string, status
...
+func (s *StorageBackendJobStore) Cleanup(ctx context.Context) error {
+	failedJobs := []string{}
+	for jobID, job := range s.InMemoryStore.GetJobs() {
+		_, err := s.storageBackend.Write(ctx, jobID, job.Output)
+
+		// Track failed jobs, log errors and continue with other jobs
+		if err != nil {
+			s.logger.ErrorContext(ctx, fmt.Sprintf("failed to persist job %s on cleanup: %s", jobID, err))
+			failedJobs = append(failedJobs, jobID)
+			continue
```
Comment (samrabelachew) [inline server/neptune/temporalworker/job/store.go:205]: don't think you need a continue here

File: server/neptune/temporalworker/job/stream_handler.go [hunk 7]
```diff
@@ -15,43 +14,51 @@ type OutputLine struct {
...
 }
 
+type StreamHandler interface {
+	Stream(jobID string, msg string)
+	Handle()
+	CloseJob(ctx context.Context, jobID string) error
+	CleanUp(ctx context.Context) error
+}
+
+// TODO: Return streaming channel
```
Comment (samrabelachew) [inline server/neptune/temporalworker/job/stream_handler.go:24]: is there anything that blocks doing this now?

File: server/neptune/temporalworker/server.go [hunk 8]
```diff
@@ -35,6 +35,12 @@ const (
 	AtlantisNamespace        = "atlantis"
 	DeployTaskqueue          = "deploy"
 	ProjectJobsViewRouteName = "project-jobs-detail"
+
+	// Equal to default terraform timeout
+	TemporalWorkerTimeout = 60 * 60 * time.Second
```
Comment (samrabelachew) [inline server/neptune/temporalworker/server.go:40]: You can just do `time.Hour`

File: server/neptune/temporalworker/server.go [hunk 9]
```diff
@@ -35,6 +35,12 @@ const (
...
+
+	// 5 minutes to allow cleaning up the job store
+	StreamHandlerTimeout = 5 * 60 * time.Second
```
Comment (samrabelachew) [inline server/neptune/temporalworker/server.go:43]: you can just do `5 * time.Minute`

File: server/neptune/temporalworker/job/stream_handler.go (hunk 7 above)
Comment (Aayyush) [reply server/neptune/temporalworker/job/stream_handler.go]: Ah, forgot to remove this comment from debugging. 

I was exploring creating the jobs output channel here and passing it to back to the caller but since we close the main channel in server, we should be creating it in the server as well. So, I chose to pass in the jobs output channel to the constructor.

File: server/neptune/temporalworker/server.go [hunk 10]
```diff
@@ -71,7 +78,8 @@ func NewServer(config *config.Config) (*Server, error) {
 	receiverRegistry := job.NewReceiverRegistry()
 
 	// terraform job output handler
-	jobStreamHandler := job.NewStreamHandler(jobStore, receiverRegistry, config.TerraformCfg.LogFilters, config.CtxLogger)
+	jobStreamChan := make(chan *job.OutputLine)
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:81]: instead of doing this, can you just expose a `Close()` or `Shutdown()` method on the handler that does this?

File: server/neptune/workflows/internal/activities/terraform_test.go [hunk 11]
```diff
@@ -28,7 +28,11 @@ func (t *testStreamHandler) Stream(jobID string, msg string) {
...
 }
 
-func (t *testStreamHandler) Close(ctx context.Context, jobID string) error {
+func (t *testStreamHandler) CloseJob(ctx context.Context, jobID string) error {
+	return nil
+}
+
+func (t *testStreamHandler) CleanUp(ctx context.Context) error {
 	return nil
 }
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/terraform_test.go:37]: what is this doing? Shouldn't this be tested? if it's not used, then the interface shouldn't have these methods.

File: server/neptune/workflows/internal/activities/job.go [hunk 12]
```diff
@@ -20,7 +20,7 @@ type CloseJobRequest struct {
 }
 
 func (t *jobActivities) CloseJob(ctx context.Context, request CloseJobRequest) error {
-	err := t.StreamCloser.Close(ctx, request.JobID)
+	err := t.StreamCloser.CloseJob(ctx, request.JobID)
 	if err != nil {
 		logger.Error(ctx, errors.Wrapf(err, "closing job").Error())
 	}
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/job.go:26]: this log doesn't really seem useful if you're returning the error.

File: server/neptune/temporalworker/server.go [hunk 13]
```diff
@@ -185,20 +196,32 @@ func (s Server) Start() error {
...
+		// has gracefully shut down
 		s.JobStreamHandler.Handle()
 	}()
 
 	<-stop
 	wg.Wait()
 
+	// Close job stream when temporalworker exits to allow gracefully shutting down the stream handler
+	s.Logger.Info("Closing job output channel")
+	close(s.JobStream)
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:210]: What happens if handle has not exited before cleanup is called?

You should be using a wait group with the handle go routine.  and reading your other response, it probably does make sense to keep the Closing of the global channel in the worker go routine given that dependence.

File: server/neptune/workflows/internal/activities/terraform_test.go (hunk 11 above)
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/terraform_test.go]: Removed the unused methods!

File: server/neptune/temporalworker/server.go (hunk 13 above)
Comment (Aayyush) [reply server/neptune/temporalworker/server.go]: Added a wait group to the handle goroutine!

File: server/neptune/workflows/internal/activities/job.go (hunk 12 above)
Comment (Aayyush) [reply server/neptune/workflows/internal/activities/job.go:26]: The reason I chose to log here was because this activity is called inside a deferred function after a `Plan` or `Apply` job is completed. Since we can't propagate the errors back up,  I decided to log the error inside the activity.
